umbrella-feeds/src/lib.rs

```diff
@@ -0,0 +1,732 @@
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<State<S>, StateApiType = S>,
+    crypto_primitives: &impl HasCryptoPrimitives,
+) -> Result<(), CustomContractError> {
+    let param: UpdateParams = ctx.parameter_cursor().get()?;
+
+    ensure!(
+        param.signers_and_signatures.len() >= host.state().required_signatures as usize,
+        CustomContractError::NotEnoughSignatures
+    );
+
+    let mut prev_signer: Option<AccountAddress> = None;
+
+    let message_hashes = contract_view_message_hash(ctx, host, crypto_primitives)?;
+
+    let required_signatures = host.state().required_signatures;
+
+    let mut validators: Vec<AccountAddress> = Vec::with_capacity(required_signatures as usize);
+
+    // To save gas we check only required number of signatures.
+    // The case, where you can have part of signatures invalid but still enough valid in total is not supported.
+    for i in 0..required_signatures {
+        let signer = param.signers_and_signatures[i as usize].0;
+        let signature = &param.signers_and_signatures[i as usize].1;
+
+        //Check signature.
+        let valid_signature =
+            host.check_account_signature(signer, signature, &message_hashes[i as usize])?;
+        ensure!(valid_signature, CustomContractError::WrongSignature);
+
+        ensure!(
+            prev_signer < Some(signer),
+            CustomContractError::SignaturesOutOfOrder
+        );
+
+        validators.push(signer);
+
+        prev_signer = Some(signer);
+    }
+
+    let are_valid_signers = host.invoke_contract_read_only::<Vec<AccountAddress>>(
+        &host.state().staking_bank,
+        &validators,
+        EntrypointName::new_unchecked("verifyValidators"),
+        Amount::zero(),
+    )?;
+
+    let are_valid_signers: bool = are_valid_signers
+        .ok_or(CustomContractError::InvokeContractError)?
+        .get()?;
+
+    ensure_eq!(are_valid_signers, true, CustomContractError::InvalidSigner);
+
+    Ok(())
+}
+
+#[receive(
+    contract = "umbrella_feeds",
+    name = "update",
+    parameter = "UpdateParams",
+    error = "CustomContractError",
+    crypto_primitives,
+    mutable
+)]
+fn update<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<State<S>, StateApiType = S>,
+    crypto_primitives: &impl HasCryptoPrimitives,
+) -> Result<(), CustomContractError> {
+    let param: UpdateParams = ctx.parameter_cursor().get()?;
+
+    let message = param.message;
+
+    // Check that the signature was intended for this contract.
+    ensure_eq!(
+        message.contract_address,
+        ctx.self_address(),
+        CustomContractError::WrongContract
+    );
+
+    // Check signature is not expired.
+    ensure!(
+        message.timestamp > ctx.metadata().slot_time(),
+        CustomContractError::Expired
+    );
+
+    // Check signature has correct chain_id.
+    ensure_eq!(
+        message.chain_id,
+        CHAIN_ID,
+        CustomContractError::ChainIdMismatch
+    );
```

Comment:
This chain_id does not seem to add value in our case. Do we really want it?

Reply:
Need @DZariusz input here. The logic/functionality is as in the original smart contracts except for a few points previously agreed on.

There is already a comment in the code. 

```
/// The concept of a CHAIN_ID does not exist on Concordium but is kept for consistency.
```

@DZariusz The reason Concordium does not need a `chain_id` for signature checks, is that the accounts that are created from your seed phrase in a wallet are different on mainnet and testnet. This means there is no risk of re-using a signature from e.g. testnet on mainnet. 
Note: The accounts on testnet/stagenet are the same when created from the same seed phrase.

Reply:
in that case we can remove this feature. I will remove it from validators code as well.

Reply:
it is removed now.